GenericList/string_list.h

```diff
@@ -1,31 +1,112 @@
 #pragma once
 
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+#include <functional>
+
 class List_String {
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
 public:
     List_String();
-    ~List_String();
 
-    int Capacity();
-    bool Capacity(int new_capacity);
-    int Count();
+    friend void swap(List_String& first, List_String& second);
+
+    //Rule of 3
+    ~List_String(); //Destructor
+    List_String(const List_String& other); //Copy Constructor
+    List_String& operator=(const List_String& other); //Copy assignement
+
+    //Rule of 5
+    //R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+    //basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+    List_String(List_String&& other);
+    List_String& operator=(List_String&& other);
+
+
+    //Remove all elements - maintain capacity
+    void Clear();
+
+    //Shrink array capacity down to count
+    void ShrinkToFit();
+
+    int Capacity() const;
+    int Count() const;
+
+    //Sets the capacity of the internal array to new_capacity. If new_capacity is smaller than Count, do nothing.
+    void Capacity(int new_capacity);
+
+    std::string ToString() const;
+
+    //Take in any amount of arguments of any type, then unpack on element construction - allows for creating new data without checking for logic errors (the compiler and the element's constructor will take care of that)
+    template<typename... Args>
+    void Add(Args&&... args) {
+        if (capacity == count) Capacity(std::max(2 * capacity, 1));
+        new (data + count++) std::string(std::forward<Args>(args)...);
+    }
+
+    std::string* Find(const std::string& val);
+    const std::string* Find(const std::string& val) const;
 
-    std::string ToString();
+    std::string* Find(std::function<bool(std::string)> pred) {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
 
-    bool Add(std::string new_val);
-    bool Contains(std::string val);
-    bool RemoveAt(int index);
+    const std::string* Find(std::function<bool(std::string)> pred) const {
+        for (auto ptr = begin(); ptr < end(); ptr++) {
+            if (pred(*ptr)) return ptr; //TODO: research why &data[i] might get overloaded
+        }
+        return nullptr;
+    }
+
+    bool Contains(const std::string& val) const;
+    void RemoveAt(int index);
+
+    int IndexOf(const std::string& val) const;
+
+    //Returns true if a relevant element exists, and removes it.
+    bool Remove(const std::string& val);
+
+    bool Remove(std::function<bool(std::string)> pred) {
```

Comment:
You should:
1. remove all elements that match the predicate instead of only the first one
2. return the number of elements removed instead of a boolean

Extra: could you make it O(N)?

Reply:
Same comment with `bool Remove(std::string const&)`